core/lib/state/src/rocksdb/mod.rs

```diff
@@ -131,23 +152,157 @@ impl RocksdbStorage {
+                &Self::serialize_state_value(value, index),
+            );
+        }
+
+        let next_key = keys
+            .last()
+            .and_then(|last_key| h256_to_u256(*last_key).checked_add(U256::one()))
+            .map(u256_to_h256);
+        match next_key {
+            None => {
+                write_batch.put_cf(
+                    StateKeeperColumnFamily::State,
+                    Self::ENUM_INDEX_MIGRATION_FINISHED,
+                    &[1u8; 1],
+                );
+                write_batch.delete_cf(
+                    StateKeeperColumnFamily::State,
+                    Self::ENUM_INDEX_MIGRATION_PROGRESS_KEY,
+                );
+                tracing::info!("RocksDB enum index migration finished");
+            }
+            Some(next_key) => {
+                write_batch.put_cf(
+                    StateKeeperColumnFamily::State,
+                    Self::ENUM_INDEX_MIGRATION_PROGRESS_KEY,
+                    next_key.as_bytes(),
+                );
+            }
+        }
+        self.db
+            .write(write_batch)
+            .expect("failed to save state data into rocksdb");
+        tracing::info!(
+            "RocksDB enum index migration chunk took {:?}, migrated {} keys",
+            started_at.elapsed(),
+            keys.len()
+        );
     }
 
     fn read_value_inner(&self, key: &StorageKey) -> Option<StorageValue> {
         let cf = StateKeeperColumnFamily::State;
         self.db
             .get_cf(cf, &Self::serialize_state_key(key))
             .expect("failed to read rocksdb state value")
-            .map(|value| H256::from_slice(&value))
+            .map(|value| H256::from_slice(&value[..32]))
     }
 
-    /// Processes storage `logs` produced by transactions.
-    fn process_transaction_logs(&mut self, updates: &HashMap<StorageKey, H256>) {
-        for (&key, &value) in updates {
-            if !value.is_zero() || self.read_value_inner(&key).is_some() {
-                self.pending_patch.state.insert(key, value);
-            }
-        }
+    fn read_enum_index(&self, key: &StorageKey) -> Option<u64> {
+        let cf = StateKeeperColumnFamily::State;
+        self.db
+            .get_cf(cf, &Self::serialize_state_key(key))
+            .expect("failed to read rocksdb state value")
+            .and_then(|value| {
+                if value.len() == 40 {
+                    Some(u64::from_be_bytes(value[32..].try_into().unwrap()))
+                } else {
+                    None
+                }
+            })
+    }
+
+    /// Returns storage logs to apply.
+    fn process_transaction_logs(
+        &self,
+        updates: HashMap<StorageKey, H256>,
+    ) -> HashMap<StorageKey, (StorageValue, Option<u64>)> {
+        updates
+            .into_iter()
+            .filter_map(|(key, value)| {
+                if self.read_value_inner(&key).is_some() {
+                    Some((key, (value, self.read_enum_index(&key))))
```

Comment:
Two reads of the same key instead of one (the second read is hot, but still). See also a comment about `StateValue` above.